### PR TITLE
feat: Migrate from Restana to Express, add large data support and misc changes

### PR DESCRIPTION
**Pull request by XavierGeerinck**
# Description

This PR introduces
* \> 4Mb data support
* Correct data serialization and content-type header setting including tests for inflation
* Invocation is now possible for non-string or object data
* Removed uuid in favor of `crypto.randomUUID()`
* Changed IRequest and IResponse from interface to type. This means that we now don't export the type by default anymore and require an import as `{ IRequest }`
* Converted `IServerImplType` and `IServerType` to a type instead of an interface
* Possibility to add your own `express` instance through the `serverHttp` option

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #330, #440, #186 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation

### CODE REVIEW

**Comment by XavierGeerinck** [comment]
Current issue, when running the test:

```javascript
const body = new Uint8Array(5 * 1024 * 1024);
const res = await server.client.invoker.invoke(daprAppId, 'test-invoker', HttpMethod.POST, body);
expect(res).toEqual(body);
```

I receive

```
"grpc: received message larger than max (31457384 vs. 10485760)"
```

Which makes no sense as the other test is passing for 10Mb 😂 but this one fails while sending 5Mb and notes 30Mb

**Comment by XavierGeerinck** [comment]
Data seems to be duplicated when serialized, is this normal behavior? 

```javascript
{"wrappers_":{"3":{"wrappers_":{"2":{"wrappers_":null,"arrayIndexOffset_":-1,"array":[null,{"type":"Buffer","data":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]}],"pivot_":1.7976931348623157e+308,"convertedPrimitiveFields_":{}},"4":{"wrappers_":null,"arrayIndexOffset_":-1,"array":[3],"pivot_":1.7976931348623157e+308,"convertedPrimitiveFields_":{}}},"arrayIndexOffset_":-1,"array":["test-invoker",[null,{"type":"Buffer","data":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]}],"application/octet-stream",[3]],"pivot_":1.7976931348623157e+308,"convertedPrimitiveFields_":{}}},"arrayIndexOffset_":-1,"array":["test-suite",null,["test-invoker",[null,{"type":"Buffer","data":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]}],"application/octet-stream",[3]]],"pivot_":1.7976931348623157e+308,"convertedPrimitiveFields_":{}}
```

File: src/types/DaprServerOptions.ts
Comment (shubham1172) [inline src/types/DaprServerOptions.ts:19]:
```suggestion
  maxBodySizeMb?: number;
```

**Comment by XavierGeerinck** [comment]
To be checked if we can merge now? @shubham1172

File: test/e2e/http/server.test.ts
Comment (shubham1172) [inline test/e2e/http/server.test.ts:143]: @XavierGeerinck do we need these?

File: src/types/DaprServerOptions.ts
Comment (shubham1172) [inline src/types/DaprServerOptions.ts:19]: Can you also add it to docs please?

**Review by shubham1172 — CHANGES_REQUESTED** [review]

File: src/types/DaprServerOptions.ts
Comment (XavierGeerinck) [reply src/types/DaprServerOptions.ts:19]: Done

File: test/e2e/http/server.test.ts
Comment (XavierGeerinck) [reply test/e2e/http/server.test.ts]: Done

**Comment by XavierGeerinck** [comment]
todo: check if test passes since it appears that HTTP might have an issue

File: src/implementation/Client/HTTPClient/HTTPClient.ts
Comment (shubham1172) [inline src/implementation/Client/HTTPClient/HTTPClient.ts:178]: All client options in debug logs?

File: src/implementation/Server/GRPCServer/GRPCServer.ts
Comment (shubham1172) [inline src/implementation/Server/GRPCServer/GRPCServer.ts:64]: Similar comments on using a default value from Settings

File: src/implementation/Server/HTTPServer/HTTPServer.ts
Comment (shubham1172) [inline src/implementation/Server/HTTPServer/HTTPServer.ts:60]: Same, use from Settings.

File: src/utils/Client.util.ts
Comment (shubham1172) [inline src/utils/Client.util.ts:112]: Shouldn't data be of type `any`?

**Review by shubham1172 — CHANGES_REQUESTED** [review]

File: src/implementation/Client/GRPCClient/GRPCClient.ts
Comment (XavierGeerinck) [reply src/implementation/Client/GRPCClient/GRPCClient.ts:78]: Hmm, I don't think we should? Else we should do that everywhere + the maxBodySize comes from the server options. If we want to do it like this I would suggest we move everything towards settings and allow users to customize the Settings util in the options parameter of the DaprClient and DaprServer

File: src/implementation/Client/HTTPClient/HTTPClient.ts
Comment (XavierGeerinck) [reply src/implementation/Client/HTTPClient/HTTPClient.ts:178]: This should be clientOptions already in the latest version?

File: src/implementation/Server/GRPCServer/GRPCServer.ts
Comment (XavierGeerinck) [reply src/implementation/Server/GRPCServer/GRPCServer.ts:64]: See previous comment

File: src/implementation/Server/HTTPServer/HTTPServer.ts
Comment (XavierGeerinck) [reply src/implementation/Server/HTTPServer/HTTPServer.ts:60]: See previous comment

File: src/implementation/Client/HTTPClient/HTTPClient.ts
Comment (shubham1172) [reply src/implementation/Client/HTTPClient/HTTPClient.ts:178]: I meant, `this.logger.debug("Fetching with clientOptions: ${JSON.stringify(clientOptions)}");` or something.

File: src/implementation/Client/GRPCClient/GRPCClient.ts
Comment (shubham1172) [reply src/implementation/Client/GRPCClient/GRPCClient.ts:78]: The idea was to have one place in the codebase where we store all the defaults. You are right, for refactoring, we could create `DaprClientOptionsDefault` and `DaprServerOptionsDefault` and use defaults when actual options are not provided. For now, I am ok either way, your call.

File: src/utils/Client.util.ts
Comment (XavierGeerinck) [reply src/utils/Client.util.ts]: I improved the serializer further and changed this to return the type

File: src/utils/Client.util.ts
Comment (XavierGeerinck) [reply src/utils/Client.util.ts]: Yes!

File: src/implementation/Client/HTTPClient/HTTPClient.ts
Comment (XavierGeerinck) [reply src/implementation/Client/HTTPClient/HTTPClient.ts:178]: Ah, it just prints the method and body and it's quite useful as it then shows what is going wrong in the execute calls

File: src/implementation/Client/GRPCClient/GRPCClient.ts
Comment (shubham1172) [reply src/implementation/Client/GRPCClient/GRPCClient.ts:78]: Tracked by #442

File: src/implementation/Server/HTTPServer/HTTPServer.ts
Comment (shubham1172) [inline src/implementation/Server/HTTPServer/HTTPServer.ts:48]: Can we add a DaprServerOption to allow users to pass their express objects? We can clearly mention that takes precedence over options like maxBodySizeMb

Comment (XavierGeerinck) [reply src/implementation/Server/HTTPServer/HTTPServer.ts:48]: I do like that idea! Will add!!

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md
Comment (shubham1172) [inline daprdocs/content/en/js-sdk-docs/js-server/_index.md:73]:
```suggestion
Instead of using the built-in web server for Dapr sidecar to application communication, you can also bring your own instance. This is helpful in scenarios like when you are building a REST API back-end and want to integrate Dapr directly in it.  

Note, this is currently available for [`express`](https://www.npmjs.com/package/express) only.
```

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md
Comment (shubham1172) [inline daprdocs/content/en/js-sdk-docs/js-server/_index.md:75]:
```suggestion
> 💡 Note: when using a custom web-server, the SDK will configure server properties like max body size, and add new routes to it. The routes are unique on their own to avoid any collisions with your application, but it's not guaranteed to not collide.
```

1. Should we configure the body size if user brings their own server? That option should be ignored in this case IMO.
2. Should we customize Dapr routes with a specific prefix? for example - `/dapr-server-<route>`, to ensure zero collisions?

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md
Comment (shubham1172) [inline daprdocs/content/en/js-sdk-docs/js-server/_index.md:81]:
```suggestion
const myApp = express();
```

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md
Comment (shubham1172) [inline daprdocs/content/en/js-sdk-docs/js-server/_index.md:83]:
```suggestion
myApp.get("/my-custom-endpoint", (req, res) => {
```

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md
Comment (shubham1172) [inline daprdocs/content/en/js-sdk-docs/js-server/_index.md:95]:
```suggestion
    serverHttp: myApp,
```

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md
Comment (shubham1172) [inline daprdocs/content/en/js-sdk-docs/js-server/_index.md:101]:
```suggestion
// Initialize subscriptions before the server starts, the Dapr sidecar uses it.
// This will also initialize the app server itself (removing the need for `app.listen` to be called).
```

File: test/unit/utils/Serializer.util.test.ts
Comment (shubham1172) [inline test/unit/utils/Serializer.util.test.ts:18]: Just a nit if you have time, these tests should be in a format - "it should ... XYZ", for e.g., `it("should serialize object to buffer as application/json")`.

File: test/e2e/http/server.test.ts
Comment (shubham1172) [inline test/e2e/http/server.test.ts:123]:
```suggestion
      const app = express();
```

File: src/types/DaprServerOptions.ts
Comment (shubham1172) [inline src/types/DaprServerOptions.ts:24]:
```suggestion
   * Express server instance.
```

**Review by shubham1172 — CHANGES_REQUESTED** [review]
@XavierGeerinck left a few thoughts.

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md
Comment (XavierGeerinck) [reply daprdocs/content/en/js-sdk-docs/js-server/_index.md:75]: Docs adaptation: OK

1. I think we should, we should still have tests passing and we require certain middleware for certain body types. Else we depend on the user to handle cloudevents for example
2. For now I would keep AS-IS, let's see if users require it first seeing the high impact of this

**Comment by XavierGeerinck** [comment]
> 

Thanks! Added the changes :) let's see if test pass

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md
Comment (shubham1172) [reply daprdocs/content/en/js-sdk-docs/js-server/_index.md]: I was just wondering if configuring max body limit on user's express instance will be intrusive, but I see your point. We should wait for some feedback on this feature and take a call I guess!

**Review by shubham1172 — APPROVED** [review]

**Review by shubham1172 — APPROVED** [review]
This is a great PR @XavierGeerinck, I can't wait for the community to try this latest addition!!

File: src/actors/ActorId.ts
Comment (DeepanshuA) [inline src/actors/ActorId.ts:14]: Should we do?
```suggestion
import { crypto } from "crypto";
```

File: src/types/http/THTTPExecuteParams.type.ts
Comment (DeepanshuA) [inline src/types/http/THTTPExecuteParams.type.ts:19]: 1. As far as I understand, BodyInit can have following types:
Blob | BufferSource | FormData | URLSearchParams | ReadableStream<Uint8Array> | string
And, object can be non-primitive type only.

It is definitely an improvement over very simplistic `string` type that we had earlier.
So, if I understand correctly, we still can't send Number etc.? - just wanted to confirm.
In Changelog, we have mentioned that 
```
and others as `application/octet-stream`
```
If my understanding is correct, Do we need to specify in docs what are the types NOT possible to be sent.

2. Instead of `BodyInit | object`, can we use `string | object` ?

Comment (XavierGeerinck) [reply src/types/http/THTTPExecuteParams.type.ts:19]: Good point, I included tests for Number and Boolean types now as well and they are passing

File: src/actors/ActorId.ts
Comment (XavierGeerinck) [reply src/actors/ActorId.ts]: It should be `import { randomUUID } from "crypto";` then, adapted accordingly :)

**Review by DeepanshuA — DISMISSED** [review]

**Comment by XavierGeerinck** [comment]
This PR will also close https://github.com/dapr/js-sdk/issues/186 due to the introduction of the httpTerminator in the correct way

**Pull request merged**
